driver-core/src/main/com/mongodb/ConnectionString.java

```diff
@@ -498,15 +500,36 @@ private void translateOptions(final Map<String, List<String>> optionsMap) {
             } else if (key.equals("sockettimeoutms")) {
                 socketTimeout = parseInteger(value, "sockettimeoutms");
             } else if (key.equals("tlsallowinvalidhostnames")) {
+                if (tlsInsecureSet) {
+                    throw new IllegalArgumentException("tlsAllowInvalidHostnames set along with tlsInsecure is not allowed");
+                }
                 sslInvalidHostnameAllowed = parseBoolean(value, "tlsAllowInvalidHostnames");
+                tlsAllowInvalidHostnamesSet = true;
             } else if (key.equals("sslinvalidhostnameallowed")) {
+                if (tlsInsecureSet) {
+                    throw new IllegalArgumentException("sslInvalidHostnameAllowed set along with tlsInsecure is not allowed");
+                }
                 sslInvalidHostnameAllowed = parseBoolean(value, "sslinvalidhostnameallowed");
+                tlsAllowInvalidHostnamesSet = true;
             } else if (key.equals("tlsinsecure")) {
+                if (tlsAllowInvalidHostnamesSet) {
```

Comment:
Could the error checking occur afterwards to DRY this up? eg afterwards check if `tlsInsecureSet` and `tlsAllowInvalidHostnamesSet` at the end and throw?

Reply:
Good idea.  I DRY'd up both this and the ssl/tls check

Reply:
Much nicer!